illwill.nim

```diff
@@ -436,9 +578,61 @@ else:  # OS X & Linux
       ord(Key.F9):        @["\e[20~"],
       ord(Key.F10):       @["\e[21~"],
       ord(Key.F11):       @["\e[23~"],
-      ord(Key.F12):       @["\e[24~"]
+      ord(Key.F12):       @["\e[24~"],
     }.toTable
 
+  proc splitInputs(inp: openarray[int], max: int): seq[seq[int]] =
+    ## splits the input buffer to extract mouse coordinates
+    var parts: seq[seq[int]] = @[]
+    var cur: seq[int] = @[]
+    for ii in inp[CSI.len+1 .. max-1]:
+      if ii == ord('M'):
+        ## Button press
+        parts.add cur
+        gMouseInfo.action = ActionPressed
+        break
+      elif ii == ord('m'):
+        ## Button release
+        parts.add cur
+        gMouseInfo.action = ActionReleased
+        break
+      elif ii != ord(';'):
+        cur.add ii
+      else:
+        parts.add cur
+        cur = @[]
+    return parts
+
+  proc getPos(inp: seq[int]): int =
+    var str = ""
+    for ii in inp:
+      str &= $(ii.chr)
+    result = parseInt(str)
+
+  proc fillGlobalMouseInfo(keyBuf: array[KeySequenceMaxLen, int]) =
+    let parts = splitInputs(keyBuf, keyBuf.len) # TODO charsRead is wrong?
+    gMouseInfo.x = parts[1].getPos() - 1
+    gMouseInfo.y = parts[2].getPos() - 1
+    let bitset = parts[0].getPos()
+    gMouseInfo.ctrl = bitset.testBit(4)
+    gMouseInfo.shift = bitset.testBit(2)
+    gMouseInfo.move = bitset.testBit(5)
+    case (bitset.uint8 shl 6 shr 6).int
+    of 0: gMouseInfo.button = MouseButton.ButtonLeft
+    of 1: gMouseInfo.button = MouseButton.ButtonMiddle
+    of 2: gMouseInfo.button = MouseButton.ButtonRight
+    else:
+      gMouseInfo.action = MouseButtonAction.ActionNone
+      gMouseInfo.button = MouseButton.ButtonNone #Left Move sends 3, but we ignore
+    gMouseInfo.scroll = bitset.testBit(6)
+    if gMouseInfo.scroll:
+      # on scroll button=3 is reported, but we want no button pressed
```

Comment:
Have you tested this on different platforms and different mouse hardware?  What about touchpads on laptops?

Reply:
i've tested scroll with touchpad and mouse (on linux, with all *normal* terminal emulators), we eventually *could* leave button=3 in cause posix is not capable of reporting more buttons! But i found it strange so removed it. Windows can report more, though.

Reply:
Ok cool.